In [1]:
import xml.etree.ElementTree as ET
tree = ET.parse('auckland_new-zealand.osm')
root = tree.getroot()

In [6]:
x =  root.findall('node')
print x.attrib

AttributeError: 'list' object has no attribute 'attrib'

In [1]:
#shorten length of file

import xml.etree.ElementTree as ET  # Use cElementTree or lxml if too slow
k = 2
OSM_FILE = "auckland_new-zealand.osm"  # Replace this with your osm file
SAMPLE_FILE = "auckland_new-zealand_every {}th.osm".format(k)

 # Parameter: take every k-th top level element

def get_element(osm_file, tags=('node', 'way', 'relation')):
    """Yield element if it is the right type of tag

    Reference:
    http://stackoverflow.com/questions/3095434/inserting-newlines-in-xml-file-generated-via-xml-etree-elementtree-in-python
    """
    context = iter(ET.iterparse(osm_file, events=('start', 'end')))
    _, root = next(context)
    for event, elem in context:
        if event == 'end' and elem.tag in tags:
            yield elem
            root.clear()


with open(SAMPLE_FILE, 'wb') as output:
    output.write('<?xml version="1.0" encoding="UTF-8"?>\n')
    output.write('<osm>\n  ')

    # Write every kth top level element
    for i, element in enumerate(get_element(OSM_FILE)):
        if i % k == 0:
            output.write(ET.tostring(element, encoding='utf-8'))

    output.write('</osm>')

In [12]:
k = 2
OSM_FILE = "auckland_new-zealand.osm"  # Replace this with your osm file
SAMPLE_FILE = "auckland_new-zealand_every {}th.osm".format(k)
file_in_use = OSM_FILE



In [48]:

import xml.etree.ElementTree as ET
tree = ET.parse(file_in_use)
root = tree.getroot()

In [5]:
from collections import defaultdict

def count_tags_attribs(filename):
    tags = defaultdict(int)
    attribs = defaultdict(int)
    for event, elem in ET.iterparse(filename):
        tags[elem.tag] += 1
        for a in elem.attrib:
            attribs[a] += 1
        
    print tags
    print '\n\n'
    print attribs
    
tags = count_tags_attribs(file_in_use)




defaultdict(<type 'int'>, {'node': 286276, 'nd': 328766, 'member': 2605, 'tag': 160300, 'relation': 138, 'way': 31473, 'osm': 1})



defaultdict(<type 'int'>, {'changeset': 317887, 'type': 2605, 'uid': 317886, 'timestamp': 317887, 'k': 160300, 'lon': 286276, 'version': 317887, 'role': 2605, 'user': 317886, 'v': 160300, 'lat': 286276, 'ref': 331371, 'id': 317887})


for the much smaller doc (doc/50) we've found these stats on tags and attribs.

tags:
defaultdict(<type 'int'>, {'node': 57256, 'nd': 64580, 'member': 667, 'tag': 31990, 'relation': 28, 'way': 6294, 'osm': 1})

attribs:
defaultdict(<type 'int'>, {'changeset': 63578, 'type': 667, 'uid': 63578, 'timestamp': 63578, 'k': 31990, 'lon': 57256, 'version': 63578, 'role': 667, 'user': 63578, 'v': 31990, 'lat': 57256, 'ref': 65247, 'id': 63578})

In [43]:
import re

lower = re.compile(r'^([a-z]|_)*$')
lower_colon = re.compile(r'^([a-z]|_)*:([a-z]|_)*$')
problemchars = re.compile(r'[=\+/&<>;\'"\?%#$@\,\. \t\r\n]')

def process_file(filename):
    keys = {"lower": 0, "lower_colon": 0, "problemchars": 0, "other": 0}
    k_attrib(file_in_use, keys)
    
    
    
def k_attrib(filename, keys):
    k_attribs = defaultdict(int)
    k_attribs_lower_colon = defaultdict(int)
    k_attribs_other = defaultdict(int)
    for event, elem in ET.iterparse(filename):
        if elem.tag == 'tag':
            k = elem.attrib['k']
            if problemchars.search(k):
                keys['problemchars'] += 1
            elif lower_colon.search(k):
                keys['lower_colon'] += 1
                k_attribs_lower_colon[k] += 1
            elif lower.search(k):
                keys['lower'] += 1
                k_attribs[k] += 1
            else:
                keys['other'] += 1
                k_attribs_other[k] += 1
    print keys
    print '\n'
    print k_attribs
    print '\n'
    print k_attribs_lower_colon
    print '\n'
    print k_attribs_other
    

    
process_file(file_in_use)
    
        
                
            
            
        

{'problemchars': 0, 'lower': 84782, 'other': 73532, 'lower_colon': 1986}


defaultdict(<type 'int'>, {'shop': 154, 'taxi': 18, 'maxspeed': 469, 'golf': 71, 'office': 3, 'restriction': 31, 'man_made': 294, 'social_facility': 2, 'gauge': 27, 'motorcycle': 7, 'seats': 2, 'trail_visibility': 4, 'network': 33, 'is_in': 25, 'toilets': 1, 'proposed': 16, 'psv': 25, 'except': 3, 'created_by': 803, 'ele': 20, 'source': 17301, 'drive_in': 1, 'wood': 163, 'location': 10, 'bollard': 2, 'usage': 23, 'covered': 15, 'population': 1, 'fax': 1, 'lit': 20, 'attribution': 18772, 'dogs': 2, 'emergency': 26, 'material': 7, 'leisure': 570, 'icao': 1, 'historic': 36, 'motor_vehicle': 36, 'foot': 312, 'resource': 1, 'tourism': 61, 'smoothness': 2, 'chimney': 4, 'campus': 3, 'colour': 6, 'fixme': 2, 'vending': 1, 'name': 3283, 'designation': 2, 'level': 16, 'circuits': 3, 'public_transport': 11, 'crossing': 19, 'leaf_cycle': 1, 'sidewalk': 13, 'kerb': 14, 'internet_access': 7, 'bicycle': 169, 'crop': 1, 'cause

From the above data we can see that anything included in the 'other' key is not useful for our observations so we will exclude them. Everything from the other two sets could be useful. 

In [41]:
def look_at_tag(tag):
    attribs = defaultdict(int)
    tags = defaultdict(int)
    x = root.findall(tag)
    for child in root:
        #y = child.findall('nd')
        for els in child:
            tags[els.tag] += 1
            for a in els.attrib:
                attribs[a] += 1

    print attribs
    print tags
    
        


look_at_tag('relation')

defaultdict(<type 'int'>, {'role': 2605, 'k': 160300, 'ref': 331371, 'type': 2605, 'v': 160300})
defaultdict(<type 'int'>, {'member': 2605, 'tag': 160300, 'nd': 328766})


from this we can see we wont have any problems with missing attributes

Node:
286276
defaultdict(<type 'int'>, {'changeset': 286276, 'uid': 286275, 'timestamp': 286276, 'lon': 286276, 'version': 286276, 'user': 286275, 'lat': 286276, 'id': 286276}

Way:
31473
defaultdict(<type 'int'>, {'changeset': 31473, 'uid': 31473, 'timestamp': 31473, 'version': 31473, 'user': 31473, 'id': 31473})
tags and attribs of children of way: defaultdict(<type 'int'>, {'role': 2605, 'k': 160300, 'ref': 331371, 'type': 2605, 'v': 160300})
defaultdict(<type 'int'>, {'member': 2605, 'tag': 160300, 'nd': 328766})

Relation:
138
defaultdict(<type 'int'>, {'changeset': 138, 'uid': 138, 'timestamp': 138, 'version': 138, 'user': 138, 'id': 138})
tags and attribs of children of relation:
defaultdict(<type 'int'>, {'role': 2605, 'k': 160300, 'ref': 331371, 'type': 2605, 'v': 160300})
defaultdict(<type 'int'>, {'member': 2605, 'tag': 160300, 'nd': 328766})

tag(child of node)
attribs in childs:defaultdict(<type 'int'>, {'role': 2605, 'k': 160300, 'ref': 331371, 'type': 2605, 'v': 160300})
attribs of 'tag':defaultdict(<type 'int'>, {'k': 160300, 'v': 160300})
tags of node:defaultdict(<type 'int'>, {'member': 2605, 'tag': 160300, 'nd': 328766})
attribs of nd: defaultdict(<type 'int'>, {'ref': 328766})

In [8]:

def find_users(filename):
    users = defaultdict(int)
    for event, elem in ET.iterparse(filename):
        try:
            users[elem.attrib['user']] += 1
        except KeyError:
            pass
    return users
        
users = find_users(file_in_use)
print len(users)
print max(users)


508
李思洋


In [14]:
"""
Your task in this exercise has two steps:

- audit the OSMFILE and change the variable 'mapping' to reflect the changes needed to fix 
    the unexpected street types to the appropriate ones in the expected list.
    You have to add mappings only for the actual problems you find in this OSMFILE,
    not a generalized solution, since that may and will depend on the particular area you are auditing.
- write the update_name function, to actually fix the street name.
    The function takes a string with street name as an argument and should return the fixed name
    We have provided a simple test so that you see what exactly is expected
"""
import xml.etree.cElementTree as ET
from collections import defaultdict
import re
import pprint

OSMFILE = file_in_use
street_type_re = re.compile(r'\b\S+\.?$', re.IGNORECASE)


expected = ["Street", "Avenue", "Boulevard", "Drive", "Court", "Place", "Square", "Lane", "Road", 
            "Trail", "Parkway", "Commons", "Crescent"]

# UPDATE THIS VARIABLE
mapping = { "St": "Street",
            "St.": "Street",
            "street": "Street",
           "Ave":"Avenue",
           
           
            }


def audit_street_type(street_types, street_name):
    m = street_type_re.search(street_name)
    if m:
        street_type = m.group()
        if street_type not in expected:
            street_types[street_type].add(street_name)


def is_street_name(elem):
    return (elem.attrib['k'] == "addr:street")


def audit(osmfile):
    osm_file = open(osmfile, "r")
    street_types = defaultdict(set)
    for event, elem in ET.iterparse(osm_file, events=("start",)):

        if elem.tag == "node" or elem.tag == "way":
            for tag in elem.iter("tag"):
                if is_street_name(tag):
                    !!audit_street_type(street_types, tag.attrib['v'])
    osm_file.close()
    return street_types



st_types = audit(OSMFILE)

pprint.pprint(dict(st_types))


{'16': set(['State Highway 16']),
 'Ave': set(['Erson Ave', 'Gillies Ave']),
 'Close': set(['Regia Close']),
 'Cr': set(['Marjorie Jayne Cr']),
 'Crescent': set(['Davis Crescent',
                  'Edenvale Crescent',
                  'Marjorie Jayne Crescent']),
 'East': set(['Customs Street East']),
 'Esplanade': set(['The Esplanade']),
 'Expressway': set(['Albany Expressway']),
 'Highway': set(['Hibiscus Coast Highway', 'Mount Wellington Highway']),
 'Howick)': set(['Fisher Parade (Howick)']),
 'Parade': set(['The Parade']),
 'Rd': set(['New North Rd']),
 'St.': set(['Queen St.']),
 'Strand': set(['The Strand']),
 'Terrace': set(['Kohia Terrace',
                 'Quest Terrace',
                 'Seaview Terrace',
                 'Taunton Terrace']),
 'Way': set(['Brian Slater Way', 'Haumia Way']),
 'West': set(['Green Lane West',
              'Sylvan Avenue West',
              'Victoria Street West']),
 'street': set(['Durham street'])}


We need to add a few street types to the expected list. I.e Terrace, Crescent. But also we need to add exceptions for streets like 'Customs Street East' or 'The Strand'. 
We'll have to check for non_capitalisation. This will be easy to fix but does add extra iterations.


In [26]:
New_expected = ['Crescent', 'Expressway', 'Terrace', 'Way', 'Close', 'Esplanade', 'Highway', 'Parade']
expected.append(New_expected)
new_maps = {
    "Cr":"Crescent"
    'Rd':'Road'
}
mapping.add(new_maps)
            
def update_name(name, mapping):
    #this has been formatted to work with the final fucntion.
    m = street_type_re.search(name)
    if m:
        street_type = m.group()
    
        if street_type in mapping:
            new_street_type = mapping[street_type]
            name = name.replace(street_type, new_street_type)
            return name
        elif street_type in expected:
            return name
        else:
            print "\n we've found an unexpected one", name
            return name
    else:
        return name
            
def change_names():
    for st_type, ways in st_types.iteritems():
            for name in ways:
                better_name = update_name(name, mapping)                



from the data we've already gatehred we can start planning how I'll parse the data into JSON format. 
There 

In [2]:
def process_map(file_in_use):
    file_out = "{0}.json".format(file_in_use)
    data = []
    with codecs.open(file_out, 'w') as fo:
        for _, element in ET.iterparse(file_in_use):
            el = shape_element(element)
            if el:
                data.append(el)


def shape_element(element):
    
    if element.tag == 'node' or element.tag == 'way':
        created = {"version":element.attrib['version'],
                  "changeset":element.attrib}
    
    correct_first_elem = {
                     "id": ""
                     "visible": ""
                     "type":""
                     "pos":""
                     "created":""
                     "changeset":""
                     "user" : ""
                     "version" : ""
                     "uid" : ""
                     "timestamp" : ""}

SyntaxError: invalid syntax (<ipython-input-2-1aff1d4b3fbd>, line 2)

In [31]:
def shape_element(element):
    node = {}
    
    if element.tag == "node" or element.tag == "way":
        created = {"version": element.attrib['version'],
                "changeset": element.attrib['changeset'],
                "timestamp":element.attrib['timestamp'],
                "user": element.attrib['user'],
                "uid": element.attrib['uid']}
        address = {}      
        node_refs = []
        try:        
            pos = [float(element.attrib['lat']), float(element.attrib['lon'])]
            node['pos'] = pos
        except KeyError:
            pass
        
        for el in element:
            if el.tag == 'nd':
                node_refs.append(el.attrib['ref'])
                
            if 'k' in el.attrib:
                k = el.attrib['k']
                val = el.attrib['v']
                if problemchars.search(k):
                    pass
                elif k.startswith('addr:'):
                    if len(k.split(':')) == 3:
                        pass
                    else:
                        k = k.split(':')[1]
                        address[k] = val
                        #need to find a way of applying street name changes
                        
                       
                else:
                    x = k.split(':')
                    if len(x) == 2:
                        if x[1] in node:
                            val_ID = node[x[1]]
                            node[x[0]] = {}
                            node[x[0]]['ID'] = val_ID
                            node[x[0]][x[1]] = val
                        
                        else:
                            node[x[0]] = {}
                            node[x[0]][x[1]] = val
                    else:
                        #len(x) = 1
                        if k in node:
                            node[k]['ID'] = val
                        else:
                            node[k] = val
        node['created'] = created
        if len(node_refs) != 0:
            node['node_refs'] = node_refs
        if len(address) != 0:
            node['address'] = address
        node['type'] = element.tag
        node['id'] = element.attrib['id']
        try:
            node['visible'] = element.attrib['visible']
        except KeyError:
            pass
                #need abettert way of doing htis
                    
                        
                        
                        
        
        """if element.attrib in CREATED:
            node['created'] = created
        if element.attrib == 'latitude' or element.attrib == 'longitude':
            
            if element.attrib[atrb]:
                node['created'][atrb] = element.attrib[atrb]
                print element.attrib[atrb]"""
        
        
                    
        # YOUR CODE HERE
        print node
        return node
    else:
        return None


def process_map(file_in, pretty = False):
    # You do not need to change this file
    file_out = "{0}.json".format(file_in)
    data = []
    with codecs.open(file_out, "w") as fo:
        for _, element in ET.iterparse(file_in):
            el = shape_element(element)
            if el:
                data.append(el)
                if pretty:
                    fo.write(json.dumps(el, indent=2)+"\n")
                else:
                    fo.write(json.dumps(el) + "\n")
    return data

In [11]:
import xml.etree.cElementTree as ET
import pprint
import re
import codecs
import json
import xml.etree.cElementTree as ET
from collections import defaultdict
import re
import pprint

k = 2
OSM_FILE = "auckland_new-zealand.osm"  # Replace this with your osm file
SAMPLE_FILE = "auckland_new-zealand_every {}th.osm".format(k)
file_in_use = OSM_FILE


street_type_re = re.compile(r'\b\S+\.?$', re.IGNORECASE)
lower = re.compile(r'^([a-z]|_)*$')
lower_colon = re.compile(r'^([a-z]|_)*:([a-z]|_)*$')
problemchars = re.compile(r'[=\+/&<>;\'"\?%#$@\,\. \t\r\n]')


expected = ["Street", "Avenue", "Boulevard", "Drive", "Court", "Place", "Square", "Lane", "Road", 'Broadway', 'Motorway', 'Parade',
            "Trail", 'Rise', "Parkway", "Commons", "Crescent", 'Crescent', 'Expressway', 'Terrace', 'Way', 'Close', 'Esplanade', 'Highway', 'Parade']

#expected ones tobe aware of: Cove, Broadway
#used this https://www.landservices.sa.gov.au/1online_services/10propertyassist/Help_for_Business_Users/StreetTypes.asp
mapping = { "St": "Street",
            "St.": "Street",
           'Dr':'Drive',
           'st' : 'Street',
            "street": "Street",
           'strret':'Street',
           'Strret':'Street',
           'road':'Road',
           "Ave":"Avenue",
           "Cr":"Crescent",
            'Rd':'Road',
           'Plc':'Place',
           'ln' : 'Lane',
           'Hwy' : 'Highway'
            }

outliers = {'East':'Street',
            #problem here with these 'we've found an unexpected one Grange Road South and South'
            'West':'Street',
            'South':'Road',
            'Strand': None, 
            'Cove': None,
             'Wharf':None,
           "School": None,
            'Tai': None,
            'Coronation' : None,
            'Subritzky' : None,
            'Redwood': None,
            'Hurstmere' : None,
            'Gardens' : None
            
           }
New_expected = []
expected.append(New_expected)

new_maps = {}
mapping.update(new_maps)

CREATED = [ "version", "changeset", "timestamp", "user", "uid"]

In [ ]:
import xml.etree.cElementTree as ET
import pprint
import re
import codecs
import json
import xml.etree.cElementTree as ET
from collections import defaultdict
import re
import pprint

k = 2
OSM_FILE = "auckland_new-zealand.osm"  # Replace this with your osm file
SAMPLE_FILE = "auckland_new-zealand_every {}th.osm".format(k)
file_in_use = OSM_FILE


street_type_re = re.compile(r'\b\S+\.?$', re.IGNORECASE)
lower = re.compile(r'^([a-z]|_)*$')
lower_colon = re.compile(r'^([a-z]|_)*:([a-z]|_)*$')
problemchars = re.compile(r'[=\+/&<>;\'"\?%#$@\,\. \t\r\n]')


expected = ["Street", "Avenue", "Boulevard", "Drive", "Court", "Place", "Square", "Lane", "Road", 'Broadway', 'Motorway', 'Parade',
            "Trail", 'Rise', "Parkway", "Commons", "Crescent", 'Crescent', 'Expressway', 'Terrace', 'Way', 'Close', 'Esplanade', 'Highway', 'Parade']

#expected ones tobe aware of: Cove, Broadway
#used this https://www.landservices.sa.gov.au/1online_services/10propertyassist/Help_for_Business_Users/StreetTypes.asp
mapping = { "St": "Street",
            "St.": "Street",
           'Dr':'Drive',
           'st' : 'Street',
            "street": "Street",
           'strret':'Street',
           'Strret':'Street',
           'road':'Road',
           "Ave":"Avenue",
           "Cr":"Crescent",
            'Rd':'Road',
           'Plc':'Place',
           'ln' : 'Lane',
           'Hwy' : 'Highway'
            }

outliers = {'East':'Street',
            #problem here with these 'we've found an unexpected one Grange Road South and South'
            'West':'Street',
            'South':'Road',
            'Strand': None, 
            'Cove': None,
             'Wharf':None,
           "School": None,
            'Tai': None,
            'Coronation' : None,
            'Subritzky' : None,
            'Redwood': None,
            'Hurstmere' : None,
            'Gardens' : None
            
           }
New_expected = []
expected.append(New_expected)

new_maps = {}
mapping.update(new_maps)

CREATED = [ "version", "changeset", "timestamp", "user", "uid"]



def update_name(name, element, node):
    #this has been formatted to work with the final fucntion.
    m = street_type_re.search(name)
    if m:
        street_type = m.group()
        if street_type in expected:
            return name, street_type
        
        elif street_type in mapping:
            new_street_type = mapping[street_type]
            name = name.replace(street_type, new_street_type)
            record_user(element, node)
            return name, new_street_type
        
        elif street_type in outliers:
            record_user(element, node)
            return name, outliers[street_type]
            #IF THE STREET TYPE MAPS TO nONE THEN WE WONT HAVE A STREET TYPE CATEGORY IN THE DATA FOR THIS STREET. good.
        else:

            name, street_type = handle_street_exceptions(name, street_type, element, node)
            record_user(element, node)
            return name, street_type

    else:
        names = name.split(' ')
        if names[-1] in expected:
            return name, names[-1]
        else:
            if name == 'Sartors Avenue':
                #for some reason this one doesn't register in the screening process
                return 'Sartors Avenue', 'Avenue'
                
            print '\n there is no street type for this one', name
            return name, None

    
def handle_street_exceptions(name, street_type, element, node):
    #check if its a state highway
    if street_type.isdigit():
        if len(name.split(' ')) > 1:
            if name.split(' ')[1] == 'Highway':
                return name, 'Highway'
            else:
                if name == '15 Arrenway Dr, Rosedale, Auckland 0632':
                    return None, None
                print '\n we have a problem with ', name
                return name, street_type
        else:
            if street_type == '26':
                return None, None
            #theres already street data for this one, including the street number
            else:
                print '\n This street name is just a number', name
                return name, street_type

    elif name == 'New Zealand':
        return None, None
    elif name == 'Sartors Avenue':
        return 'Sartors Avenue', 'Avenue'
    elif name == 'Waipuna Road, Mt Wellington':
        names = name.split(' ')
        names[2] = ''
        names[3] = ''
        names[1] = names[1].strip(',')
        name = ''
        name = name.join(names)
        name, street_type = update_name(name, element, node)
    elif name == 'Exmouth Road, Northcote, Auckland':
        return None, None
    elif name == 'Cnr St Lukes Rd & Wagener Plc,':
        name = name.replace('Cnr St Lukes Rd & ', '')
        name = name.strip(',')
        name, street_type = update_name(name, element, node)
        return name, street_type
    elif name == 'Gillies Avenue, Newmarket':
        name = name.replace(', Newmarket', '')
        name, street_type = update_name(name, element, node)
        return name, street_type
    elif street_type == 'Howick)':
        name = name.replace(' (Howick)', '')
        name, street_type = update_name(name, element, node)
        return name, street_type
    elif street_type == 'beach':
        if name == 'Little Oneroa beach':
            name == 'Little Oneroa Beach'
            return name, None
        else:
            
            street_type = None
            names = name.split(' ')
            names[-1] = names[-1].capitalize()
            name = ""
            for n in names:
                name + n + ' '
            print name, street_type, 'xx'
            return name, street_type
    elif name == 'Victoria St W':
        names = name.split(' ')
        names[1] = mapping[names[1]]
        names[2] = 'West'
        name = names[0] + ' ' + names[1] + ' ' + names[2]
        return name, outliers[names[2]]
    elif name == 'WaipunaRoad':
        return 'Waipuna Road', 'Road'
        
            
        
    else:
        print "\n we've found an unexpected one", name, 'and', street_type
        return name, street_type
    
    
def handle_attribute_exceptions(k, v):
    print 'hi;'
    
def record_user(element, node):
    #used to recxord users who dont follow the expected styles of dta entry. This is very rudimentary so if you're one of these 'users' dont be offended.
    if 'bad_user' in node:
        pass
    else:
        if 'user' in node['created'] and 'uid' in node['created']:
            node['bad_user'] = [node['created']['user'], node['created']['uid']]
        elif 'user' in node['created']:
            node['bad_user'] = node['created']['user']
        elif 'uid' in node['created']:
            node['bad_user'] =  node['created']['uid']
            

def shape_element(element):
    #this will be the function which does themost work
    #we'll only include the node and way tags. 
    
    node = {}
    if element.tag == "node" or element.tag == "way":
        #go through all the initial attribs 
        created = {}
        for c in CREATED:
            try:
                created[c] = element.attrib[c]
            except KeyError:
                pass
            
        node['created'] = created
        try:        
            pos = [float(element.attrib['lat']), float(element.attrib['lon'])]
            node['pos'] = pos
        except KeyError:
            pass
        try:
            node['visible'] = element.attrib['visible']
        except KeyError:
            pass
        node['type'] = element.tag
        node['id'] = element.attrib['id']
        
        
        
        
        address = {}      
        node_refs = []
        testo = True
        param = False
        if testo: 
            for child in element:
                if child.tag == 'nd':
                    node_refs.append(child.attrib['ref'])

                if 'k' in child.attrib:
                    k = child.attrib['k']
                    v = child.attrib['v']
                    if problemchars.search(k):
                        record_user(element, node)
                        #dont think this will happen at all but just in case
                        pass
                    
                    elif k.startswith('addr:'):
                        if len(k.split(':')) == 3:
                            print '\n look at this!', k
                            #this is too long too bother with
                            
                        else:
                            #find the sub address bit
                            k = k.split(':')[1]
                            if k == 'city':
                                k = 'area'
                                address[k] = v
                                #not clear enough what these values for 'city' represented so changed the category name
                            if k == 'street':
                                street_type = None
                                
                                try:              
                                    v, street_type = update_name(v, element, node)
                                except TypeError:
                                    if v == 'Waipuna Road, Mt Wellington':
                                        v = 'Waipuna Road'
                                        street_type = 'Road'
                                        address['area'] = 'Mt Wellington'
                                        
                                    else:    
                                    #have a look at this Waipuna Road, Mt Wellington street
                                        print 'have a look at this', v, k
                                if v:
                                    address['street'] = v
                                if street_type:
                                    address['street_type'] = street_type
                                if v == '15 Arrenway Dr, Rosedale, Auckland 0632':
                                    v, street_type = update_name('Arrenway Dr')
                                    address['street'] = v
                                    address['housenumber'] = '15'
                                    address['street_type'] = street_type
                                    address['area'] = 'Rosedale'
                                    address['postcode'] = '0632'
                                if v == 'Exmouth Road, Northcote, Auckland':
                                    v, street_type = update_name('Exmouth Road')
                                    address['street'] = v
                                    address['street_type'] = street_type
                                    address['area'] = 'Northcote'
                                    
                                    
                    else:

                        
                        #k will be the val fpor a restaurant or something
                        x = k.split(':')
                        if len(x) == 2:
                            category = x[0]
                            sub_cat = x[1]
                            #problem here is when we have single values which arent dicts. 
                            
                            if category == 'LINZ':
                                pass
                            elif category in node:
                                if type(node[category]) == dict:
                                    node[category][sub_cat] = v
                            
                                else:
                                    #category in node:
                                    category_id = node[category]
                                    node[category] = {}
                                    node[category]['ID'] = category_id
                                    node[category][sub_cat] = v
                            else:                                  
                                node[category] = {}
                                node[category][sub_cat] = v
                        elif len(x) == 1:
                            #assuming the rest will be
                            if k in node:
                                if k == 'type':
                                    pass
                                else:
                                    if type(node[k]) == dict:
                                        print 'this doesnt happen very often1', k, node[k]
                                        print element.tag, element.attrib.items()
                                        for c in element:
                                            print c.attrib.items(), c.tag
                                        #id say we'd have to put numbered keys down
                                    else:
                                        
                                        #id say we'd have to put numbered keys down
                                        if node[k] == '309844804':
                                            node = None
                                            return node
                                            
                                            #this was a surveying entry with entries like [('k', 'calc_area'), ('v', '47251')], and 
                                            #it had an extra ID value wchich was causing errors. 
                                            #so i decided the most efficient thing would be just to remove the node alltogether
                                         
                                        print 'this doesnt happen very often2', k, node[k]
                                        print element.tag, element.attrib.items()
                                        for c in element:
                                            print c.attrib.items(), c.tag
                            else:
                                node[k] = v
                                
                                
        if len(node_refs) != 0:
            node['node_refs'] = node_refs
        if len(address) != 0:
            node['address'] = address
            
        return node
    else:
        return None

                    
                        
                        

def process_map(file_in, pretty = False):
    # You do not need to change this file
    file_out = "{0}.json".format(file_in)
    data = []
    count = 0
    with codecs.open(file_out, "w") as fo:
        for _, element in ET.iterparse(file_in):
            count += 1
            if count % 100000 == 0:
                print count
            el = shape_element(element)
            if el:
                data.append(el)
                if pretty:
                    fo.write(json.dumps(el, indent=2)+"\n")
                else:
                    fo.write(json.dumps(el) + "\n")
    return data

def test():
    # NOTE: if you are running this code on your computer, with a larger dataset, 
    # call the process_map procedure with pretty=False. The pretty=True option adds 
    # additional spaces to the output, making it significantly larger.
    data = process_map(file_in_use)
    #pprint.pprint(data)
    return data

data = test()
print 'done'







Problems I came across in the processing step:
-Inconsistent address recording and recording style.
-Inconsistent area category naming. (some areas are recorded to a good degree of accuracy, others refer to a large area in auckland such as 'Auckland' which is more of a general area in Auckland which includes a lot of the most central areas of the city.)


The file processing method involved processing the address's for each Node/Way tag. Doing this, I found a fair few inconsistencies. A lot could be cleaned up using similar code to that used in 'Improving Street Names' from the Case Study quiz, but there was a proportion which I had to clean up almost manually. They would have multiple problems, one of them that came up often was some street address's were recorded in the same line as their postcode, city area and other address categories. I had to sort these and put them into their proper categories for the JSON document. For example 'Exmouth Road, Auckland' has the 'Area' category included in it (Auckland)(This is something probably only noticed by a New Zealander - Auckland being the central area of Auckland). Problems like this have had to be raised during processing and exceptions/solutions written in the 'handle_street_exceptions' function to fix them. 
While going through exceptions to my original code and updating my functions to handle those exceptions I used the list of street types in this website to differentiate between a valid street type or an instance where I'd leave out the street type detail. (https://www.landservices.sa.gov.au/1online_services/10propertyassist/Help_for_Business_Users/StreetTypes.asp).

As for the inconsistent area naming, I don't yet know how to write a process to automatically map street addresses to city areas using external data. Perhaps I'll learn this later in the course?! 
- I feel the area naming is still good enough to make observations during the query process.

I wonder if specific users are responsible for the bad address naming conventions and if so a faster method of processig the data is through quickly identifying and then looking for those users and then their bad data. This would save having to trawl through all of the data points individually usign my long iterative screening process for bad data.
- Note on this, I've included data points for all those street addresses where I've had to make changes to them.

My processing didn't involve much more than what I've explained. I was processing the files with the intention of looking at the address data so I didn't worry about any other potenital inconsistencies in the data. 





For example, the Street name which is found in a tag, tag, comes under the 'k' attribute 'address:street' and the 'v' attribute 'Cracroft Strret' is processed so in the JSON document there will be now a node/way with an address category like so:
'address': {'street':'Carcroft Street', 'street_type':'Street'}    - There could also be extra details in the address like Postcode and Housenumber.




In [95]:
for i in data:
    if 'address' in i:
        print i['address'][]

{'amenity': 'library', 'name': 'Kumeu Library', 'created': {'uid': '10829', 'changeset': '38742733', 'version': '4', 'user': 'myfanwy', 'timestamp': '2016-04-21T00:25:30Z'}, 'pos': [-36.7722439, 174.5476509], 'address': {'street': 'Main Road', 'street_type': 'Road', 'area': 'Huapai, Kumeu'}, 'type': 'node', 'id': '833756315'}
{'shop': 'dry_cleaning', 'name': 'T&C Drycleaners', 'created': {'uid': '252775', 'changeset': '6643685', 'version': '1', 'user': 'LINZ Data Upload', 'timestamp': '2010-12-13T00:28:05Z'}, 'pos': [-36.8540555, 174.7458883], 'address': {'street': 'Ponsonby Road', 'street_type': 'Road', 'area': 'Auckland'}, 'type': 'node', 'id': '1036654014'}
{'shop': 'clothes', 'name': 'Kathmandu', 'created': {'uid': '70696', 'changeset': '7787175', 'version': '3', 'user': 'xybot', 'timestamp': '2011-04-06T17:39:29Z'}, 'pos': [-36.8827273, 174.6358369], 'address': {'street': 'Vitasovich Avenue', 'street_type': 'Avenue'}, 'type': 'node', 'id': '1234128432'}
{'id': '1668246187', 'addre

issues:
'created_by': 'Merkaartor 0.12'
seems likie rossB is upto no good
'created_by': 'Potlatch 0.5
'myfanwy'



'LINZ': {'layer': 'descrip_text', 'point_size': '6.0', 'text_placement_justification': '1.0', 'y_offset': '1.0', 'x_offset': '-1.0', 'source_version': 'V16', 'merged': 'V16', 'size': '4.1'},
more linz shit(showed up when testing for problemchars:
 error chars at nzopengps.org:node_id 
 items: [('k', 'nzopengps.org:node_id'), ('v', '10783')] 
 tag tag
node
tag xxx [('k', 'LINZ:dataset'), ('v', 'nzopengps_auckland')]
tag xxx [('k', 'nzopengps.org:node_id'), ('v', '10783')]
tag xxx [('k', 'nzopengps.org:source_version'), ('v', 'r4999')]

 error chars at nzopengps.org:source_version 
 items: [('k', 'nzopengps.org:source_version'), ('v', 'r4999')] 
 tag tag
node
tag xxx [('k', 'LINZ:dataset'), ('v', 'nzopengps_auckland')]
tag xxx [('k', 'nzopengps.org:node_id'), ('v', '10783')]
tag xxx [('k', 'nzopengps.org:source_version'), ('v', 'r4999')]




{'comment': 'Determined via KeypadMapper. Feel free to correct the location'




where street type is broadway we have four:
<Element 'tag' at 0x00000000854D3E40> tag [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'name'), ('v', 'Rialto Newmarket')]
tag xxx [('k', 'phone'), ('v', '+64 9 369 2417')]
tag xxx [('k', 'amenity'), ('v', 'cinema')]
tag xxx [('k', 'website'), ('v', 'http://www.rialto.co.nz/Cinema/Newmarket')]
tag xxx [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'addr:housenumber'), ('v', '167-169')]

 we've found an unexpected one Broadway and Broadway

<Element 'tag' at 0x00000000853CF360> tag [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'name'), ('v', 'The Claddagh')]
tag xxx [('k', 'phone'), ('v', '+64 9 522 4410')]
tag xxx [('k', 'amenity'), ('v', 'pub')]
tag xxx [('k', 'website'), ('v', 'http://www.thecladdagh.co.nz/')]
tag xxx [('k', 'addr:city'), ('v', 'Auckland')]
tag xxx [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'addr:housenumber'), ('v', '372')]

 we've found an unexpected one Broadway and Broadway

<Element 'tag' at 0x00000000913708D0> tag [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'name'), ('v', 'Event Cinemas')]
tag xxx [('k', 'phone'), ('v', '+64 9 369 2441')]
tag xxx [('k', 'amenity'), ('v', 'cinema')]
tag xxx [('k', 'website'), ('v', 'http://www.eventcinemas.co.nz/Cinema/Broadway')]
tag xxx [('k', 'operator'), ('v', 'Skycity')]
tag xxx [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'addr:housenumber'), ('v', '77')]

 we've found an unexpected one Broadway and Broadway

<Element 'tag' at 0x0000000091370210> tag [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'name'), ('v', 'The Olympic')]
tag xxx [('k', 'phone'), ('v', '+64 9 522 4414')]
tag xxx [('k', 'sport'), ('v', 'swimming')]
tag xxx [('k', 'leisure'), ('v', 'swimming_pool')]
tag xxx [('k', 'website'), ('v', 'http://www.theolympic.co.nz/')]
tag xxx [('k', 'addr:city'), ('v', 'Auckland')]
tag xxx [('k', 'addr:street'), ('v', 'Broadway')]
tag xxx [('k', 'addr:housenumber'), ('v', '77')]

-These are the .items() involved with the tag. Should probably include them as an example of my cleaning. theyre all the children of node tags. 



<Element 'tag' at 0x000000005901C600> tag [('k', 'addr:street'), ('v', 'New Zealand')]
node
tag xxx [('k', 'name'), ('v', 'Handmade Burgers')]
tag xxx [('k', 'phone'), ('v', '8494590')]
tag xxx [('k', 'amenity'), ('v', 'fast_food')]
tag xxx [('k', 'addr:street'), ('v', 'New Zealand')]
tag xxx [('k', 'addr:housenumber'), ('v', '455')]



<Element 'tag' at 0x0000000064E3AE70> tag [('k', 'addr:street'), ('v', 'Clearwater Cove')]
node
tag xxx [('k', 'name'), ('v', 'The Flagship Marina Restaurant and Bar')]
tag xxx [('k', 'amenity'), ('v', 'restaurant')]
tag xxx [('k', 'addr:city'), ('v', 'Auckland')]
tag xxx [('k', 'addr:street'), ('v', 'Clearwater Cove')]
tag xxx [('k', 'addr:postcode'), ('v', '0618')]
tag xxx [('k', 'addr:housenumber'), ('v', '15')]


We found the one with(Howick) as a street type. need to look at that again.



In [84]:
node [('changeset', '5468561'), ('uid', '142831'), ('timestamp', '2010-08-12T04:18:42Z'), ('lon', '174.9323109'), ('version', '1'), ('user', 'Rudy355'), ('lat', '-37.0350831'), ('id', '856116245')]
<built-in method items of dict object at 0x000000001B143378> tag

 This street name is just a number 26
node [('changeset', '17338787'), ('uid', '1713475'), ('timestamp', '2013-08-14T00:09:24Z'), ('lon', '174.7739061'), ('version', '3'), ('user', 'Maryw221'), ('lat', '-36.8709151'), ('id', '2418374878')]
<built-in method items of dict object at 0x000000004CBCD9D8> tag
<built-in method items of dict object at 0x000000004CBCDAE8> tag
<built-in method items of dict object at 0x000000004CBCDD08> tag
<built-in method items of dict object at 0x000000004CBCDE18> tag
<built-in method items of dict object at 0x000000004CBD0048> tag
<built-in method items of dict object at 0x000000004CBD0158> tag
<built-in method items of dict object at 0x000000004CBD0268> tag
<built-in method items of dict object at 0x000000004CBD0378> tag
<built-in method items of dict object at 0x000000004CBD0488> tag
<built-in method items of dict object at 0x000000004CBD0598> tag
<built-in method items of dict object at 0x000000004CBD06A8> tag

 we've found an unexpected one Gillies Avenue, Newmarket and Newmarket

 we've found an unexpected one Red Beach Primary School and School

 we've found an unexpected one Victoria St W and W

 we've found an unexpected one Auckland Southern Motorway and Motorway

 we've found an unexpected one Cracroft Strret and Strret

 we've found an unexpected one Cnr St Lukes Rd & Wagener Plc, and Plc,

 we've found an unexpected one Coronation and Coronation

 we've found an unexpected one Ara Tai and Tai
this is howiucks dets Fisher Parade (Howick)

 we have a problem with  15 Arrenway Dr, Rosedale, Auckland 0632

 we've found an unexpected one Waipuna Road, Mt Wellington and Wellington

 we've found an unexpected one Bampton Rise and Rise

 there is no street type for this one Sartors Avenue 

 we've found an unexpected one Queen st and st

 we've found an unexpected one Pacific Rise and Rise
this doesnt happen very often id 309844804
done

In [98]:
int('j')

ValueError: invalid literal for int() with base 10: 'j'

In [99]:
print 'it' - 't'

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [108]:
we've found an unexpected one Gillies Avenue, Newmarket and Newmarket






 we've found an unexpected one Cnr St Lukes Rd & Wagener Plc, and Plc,

 we've found an unexpected one  and Coronation


this is howiucks dets 

 we have a problem with  15 Arrenway Dr, Rosedale, Auckland 0632

 we've found an unexpected one Waipuna Road, Mt Wellington and Wellington







this doesnt happen very often id 309844804
done

AttributeError: 'list' object has no attribute 'items'

In [20]:

    
update_name('Pacific Rise')
    

('Pacific Rise', 'Rise')

In [7]:
x = 'g' +'h'

In [8]:
we've caught one
node [('changeset', '17338787'), ('uid', '1713475'), ('timestamp', '2013-08-14T00:09:24Z'), ('lon', '174.7739061'), ('version', '3'), ('user', 'Maryw221'), ('lat', '-36.8709151'), ('id', '2418374878')]
[('k', 'name'), ('v', 'PINNACLE LIFE NZ')] tag
[('k', 'note'), ('v', 'Established in 1998 with offices in Newmarket, Auckland, Pinnacle Life Limited is a New Zealand owned life insurer and member of the FSC, offering insurance products directly to consumers. Visit PINNACLE LIFE website for more information.')] tag
[('k', 'phone'), ('v', '+64 9 522 5515')] tag
[('k', 'website'), ('v', 'https://www.pinnaclelife.co.nz')] tag
[('k', 'building'), ('v', 'office')] tag
[('k', 'addr:city'), ('v', 'Auckland')] tag
[('k', 'wheelchair'), ('v', 'yes')] tag
[('k', 'addr:street'), ('v', 'Gillies Avenue, Newmarket')] tag
[('k', 'addr:postcode'), ('v', '1023')] tag
[('k', 'addr:housename'), ('v', 'Suit 1, Level 2, Gillies Avenue Business Park')] tag
[('k', 'addr:housenumber'), ('v', '27')] tag

 we've found an unexpected one Gillies Avenue, Newmarket and Newmarket
this is that beach Little Oneroa beach beach
Little Oneroa beach beach
Little Oneroa beach
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-36-6b9e14e5b87f> in <module>()
    273     return data
    274 
--> 275 data = test()
    276 print 'done'

<ipython-input-36-6b9e14e5b87f> in test()
    269     # call the process_map procedure with pretty=False. The pretty=True option adds
    270     # additional spaces to the output, making it significantly larger.
--> 271     data = process_map(file_in_use)
    272     #pprint.pprint(data)
    273     return data

<ipython-input-36-6b9e14e5b87f> in process_map(file_in, pretty)
    256     with codecs.open(file_out, "w") as fo:
    257         for _, element in ET.iterparse(file_in):
--> 258             el = shape_element(element)
    259             if el:
    260                 data.append(el)

<ipython-input-36-6b9e14e5b87f> in shape_element(element)
    179                                 if v:
    180                                     address['street'] = v
--> 181                                 if street_type:
    182                                     address['street_type'] = street_type
    183                                 if v == '15 Arrenway Dr, Rosedale, Auckland 0632':

UnboundLocalError: local variable 'street_type' referenced before assignment

'gh'

In [ ]:
def update_name(name):
    #this has been formatted to work with the final fucntion.
    m = street_type_re.search(name)
    if m:
        street_type = m.group()
        if street_type in expected:
            return name, street_type
        
        elif street_type in mapping:
            new_street_type = mapping[street_type]
            name = name.replace(street_type, new_street_type)
            return name, new_street_type
        
        elif street_type in outliers:
            return name, outliers[street_type]
            #IF THE STREET TYPE MAPS TO nONE THEN WE WONT HAVE A STREET TYPE CATEGORY IN THE DATA FOR THIS STREET. good.
        else:
            try:
                name, street_type = handle_street_exceptions(name, street_type)
                return name, street_type
            except TypeError:
                print name, street_type
    else:
        names = name.split(' ')
        if names[-1] in expected:
            return name, names[-1]
        print '\n there is no street type for this one', name
        return name, None

    
def handle_street_exceptions(name, street_type):
    #check if its a state highway
    if street_type.isdigit():
        if len(name.split(' ')) > 1:
            if name.split(' ')[1] == 'Highway':
                return name, 'Highway'
            else:
                print '\n we have a problem with ', name
                return name, street_type
        else:
            if street_type == '26':
                return None, None
            #theres already street data for this one, including the street number
            else:
                print '\n This street name is just a number', name
                return name, street_type

    elif name == 'New Zealand':
        return None, None
    elif name == 'Waipuna Road, Mt Wellington':
        names = name.split(' ')
        names[2] = ''
        names[3] = ''
        names[1] = names[1].strip(',')
        name = ''
        name = name.join(names)
        name, street_type = update_name(name)
    elif name == '15 Arrenway Dr, Rosedale, Auckland 0632':
        return None, None
    elif name == 'Cnr St Lukes Rd & Wagener Plc':
        name = name.replace('Cnr St Lukes Rd & ', '')
        name, street_type = update_name(name)
        return name, street_type
    elif name == ' Gillies Avenue, Newmarket':
        name = name.replace(', Newmarket', '')
        name, street_type = update_name(name)
        return name, street_type
    elif street_type == 'Howick)':
        name = name.replace(' (Howick)', '')
        name, street_type = update_name(name)
        return name, street_type
    elif street_type == 'beach':
        street_type = 'Beach'
        names = name.split(' ')
        names[-1] = names[-1].capitalize()
        name = ''
        name = name.join(names)
        print name, street_type
        return name, street_type
    elif name == 'Victoria St W':
        names = name.split(' ')
        names[1] = mapping[names[1]]
        names[2] = 'West'
        name = names[0] + ' ' + names[1] + ' ' + names[2]
        return name, outliers[names[2]]
    elif name == 'WaipunaRoad':
        return 'Waipuna Road', 'Road'
        
            
        
    else:
        print "\n we've found an unexpected one", name, 'and', street_type
        return name, street_type

In [2]:
print len(None)

TypeError: object of type 'NoneType' has no len()

In [10]:
print 200%100

0
